python/ray/autoscaler/node_provider.py

```diff
@@ -33,13 +30,148 @@ def __init__(self, provider_config: Dict[str, Any], cluster_name: str) -> None:
         """
         return False
 
+    def create_node(
+        self, node_config: Dict[str, Any], tags: Dict[str, str], count: int
+    ) -> Optional[Dict[str, Any]]:
+        """Creates a number of nodes within the namespace.
+
+        Args:
+            node_config: the "node_config" section of specific node type (under
+                "available_node_types" section) in the autoscaler config yaml.
+                The node type is decided by the node launcher.
+            tags: the tags to be set to the created nodes
+            count: the number of nodes to be created
+
+        Optionally returns a mapping from created node ids to node metadata.
+        The return value is not used by the autoscaler, but may be useful for debugging.
+        """
+        raise NotImplementedError
+
+    def create_node_with_resources(
+        self,
+        node_config: Dict[str, Any],
+        tags: Dict[str, str],
+        count: int,
+        resources: Dict[str, float],
+    ) -> Optional[Dict[str, Any]]:
+        """Create nodes with a given resource config.
+
+        Ignore this function for now---simply forward the call to create_node()
+        """
+        return self.create_node(node_config, tags, count)
+
+    def get_command_runner(
+        self,
+        log_prefix: str,
+        node_id: str,
+        auth_config: Dict[str, Any],
+        cluster_name: str,
+        process_runner: ModuleType,
+        use_internal_ip: bool,
+        docker_config: Optional[Dict[str, Any]] = None,
+    ) -> CommandRunnerInterface:
+        """Returns the CommandRunner class used to perform SSH commands.
+
+        Args:
+            log_prefix: stores "NodeUpdater: {}: ".format(<node_id>). Used
+                to print progress in the CommandRunner.
+            node_id: the node ID.
+            auth_config: the authentication configs from the autoscaler
+                yaml file.
+            cluster_name: the name of the cluster.
+            process_runner: the module to use to run the commands
+                in the CommandRunner. E.g., subprocess.
+            use_internal_ip: whether the node_id belongs to an internal ip
+                or external ip.
+            docker_config: If set, the docker information of the docker
+                container that commands should be run on.
+        """
+        common_args = {
+            "log_prefix": log_prefix,
+            "node_id": node_id,
+            "provider": self,
+            "auth_config": auth_config,
+            "cluster_name": cluster_name,
+            "process_runner": process_runner,
+            "use_internal_ip": use_internal_ip,
+        }
+        if docker_config and docker_config["container_name"] != "":
+            return DockerCommandRunner(docker_config, **common_args)
+        else:
+            return SSHCommandRunner(**common_args)
+
+    def terminate_node(self, node_id: str) -> Optional[Dict[str, Any]]:
```

Comment:
Looks like some of the methods were moved around. Would you mind moving them back to reduce the diff?

Reply:
Hi! So I tried to classify the APIs into 3 categories at the beginning of the document, and changed node_provider.py to follow the logical orders in the doc. I could change it back if you prefer the original one.

Reply:
That makes sense.

However, to make the review easier, it would be better to do the reordering in its own PR (to be merged either before or after this one.)

Reply:
OK I have moved it back. Let's see whether we need another PR after this one